### PR TITLE
On chain RHS publishing

### PR DESCRIPTION
**Pull request by Kolezhniuk**

### CODE REVIEW

File: src/credentials/credential-wallet.ts [hunk 1]
```diff
@@ -367,6 +367,20 @@ export class CredentialWallet implements ICredentialWallet {
...
+      case CredentialStatusType.Iden3OnchainSparseMerkleTreeProof2023: {
+        const issuerId = DID.idFromDID(issuer);
+        const chainId = getChainId(DID.blockchainFromId(issuerId), DID.networkIdFromId(issuerId));
+        return {
+          ...credentialStatus,
+          // `[did]:[methodid]:[chain]:[network]:[id]/credentialStatus?(revocationNonce=value)&[contractAddress=[chainID]:[contractAddress]]&(state=issuerState)`
+
+          id: `${issuer.string()}/credentialStatus?revocationNonce=${
+            request.revocationOpts.nonce || ''
+          }&contractAddress=${chainId}:${request.revocationOpts.id}&state=${
```
Comment (vmidyllic) [inline src/credentials/credential-wallet.ts:379]: I would not include &state or &revocationNonce at all if corresponding values are empty

**Review by vmidyllic — APPROVED** [review]

**Review by volodymyr-basiuk — APPROVED** [review]

**Review by vmidyllic — APPROVED** [review]

File: src/credentials/status/credential-status-publisher.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+   */
+  public get(type: CredentialStatusType): ICredentialStatusPublisher[] | undefined {
+    return this._publishers.get(type);
+  }
+}
+
+/**
+ * Implementation of the ICredentialStatusPublisher interface for publishing on-chain credential status.
+ */
+export class OnChainCredentialStatusPublisher implements ICredentialStatusPublisher {
```
Comment (vmidyllic) [inline src/credentials/status/credential-status-publisher.ts:49]: rename to Iden3OnchainSmtCredentialStatusPublisher

File: src/credentials/status/credential-status-publisher.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    await txPromise;
+  }
+}
+
+/**
+ * Implementation of the ICredentialStatusPublisher interface for publishing off-chain credential status.
+ */
+export class OffChainCredentialStatusPublisher implements ICredentialStatusPublisher {
```
Comment (vmidyllic) [inline src/credentials/status/credential-status-publisher.ts:87]: rename to Iden3SmtRhsCredentialStatusPublisher

File: src/storage/blockchain/onchain-revocation.ts [hunk 4]
```diff
@@ -70,6 +75,46 @@ export class OnChainRevocationStorage {
     };
   }
 
+  public async saveNodes(payload: bigint[][]): Promise<TransactionReceipt> {
```
Comment (vmidyllic) [inline src/storage/blockchain/onchain-revocation.ts:78]: maybe this should be publisher logic?  not sure

Comment (vmidyllic) [reply src/storage/blockchain/onchain-revocation.ts:78]: ok let's keep it for now

File: src/identity/identity-wallet.ts [hunk 5]
```diff
@@ -858,16 +909,53 @@ export class IdentityWallet implements IIdentityWallet {
     );
   }
 
-  private getKMSIdByAuthCredential(credential: W3CCredential): KmsKeyId {
-    if (credential.type.indexOf('AuthBJJCredential') === -1) {
-      throw new Error("can't sign with not AuthBJJCredential credential");
+  /** {@inheritDoc IIdentityWallet.publishStateToReverseHashService} */
+  async publishStateToReverseHashService(opts: RHSOptions): Promise<void> {
+    if (!opts.issuerDID && !opts.treeModel) {
```
Comment (vmidyllic) [inline src/identity/identity-wallet.ts:914]: issuerDID should be mandatory.
decisison is based on treeModel
publishStateToReverseHashService  is  publishToCredentialRegistry

File: src/credentials/rhs.ts [hunk 6]
```diff
@@ -23,12 +20,27 @@ export interface TreesModel {
   state: Hash;
 }
 
+/**
+ * Options for RhsCommon.
+ */
+export type RHSOptions = {
```
Comment (vmidyllic) [inline src/credentials/rhs.ts:26]: move to wallet, rename

**Review by vmidyllic — APPROVED** [review]

**Review by vmidyllic — APPROVED** [review]

**Pull request merged**